Modify tag list

almgong opened the pull request:
Closes #14

wilsonyqm commented on app/controllers/images_controller.rb:
```diff
@@ -50,6 +50,14 @@ def destroy
   end
 
+  def edit
+    @image = Image.find(params[:id])
```
it's better to use `find_by` here, `find` does assertion and validation, `find_by` can help you catch the error and deal with it.

almgong replied:
Sure I can update them to use `find_by`

wilsonyqm commented on app/views/images/edit.html.erb:
```diff
@@ -0,0 +1,16 @@
+        <%= image_tag(@image.url) %>
+      </div>
+      <div class="form-group">
+        <%= form.input :tag_list, label: 'Tags for your image (comma separated)', input_html: { class: 'form-control', value: @image.tag_list.to_s } %>
```
Can you organize each hash into separate rows, just to make it clear

almgong replied:
Sure

wilsonyqm commented on app/views/images/edit.html.erb:
```diff
@@ -0,0 +1,16 @@
+  <div class="col-sm-6 col-sm-offset-3 mx-auto">
+    <h2>Edit Image:</h2>
+    <%= simple_form_for(@image) do |form| %>
+      <div class="form-group">
```
what is this class used for?

almgong replied:
`form-group` is used just for styling the form elements. Aaron is still in the process of reviewing a follow up PR for styles, where this class is no longer needed since we use simple form with bootstrap integration. Once that PR is merged, I will go and update any forms with this extra class.

wilsonyqm commented on app/views/images/index.html.erb:
```diff
@@ -10,7 +10,7 @@
             </p>
             <a href="<%= image_path image %>" class="btn btn-primary">See image</a>
-            <a href="#" class='btn btn-secondary' data-toggle="modal" data-target="#shareModal" data-form-action="<%= image_share_index_path(image.id) %>" data-image-url="<%= image.url %>">Share!</a>
+            <%= render partial: 'image_shared_actions', locals: { image: image } %>
```
actually in each partial, you can directly use `@image`, and don't need to pass it as locals.

almgong replied:
For the index, there isn't an instance variable `@image` to pass in, since we are rendering this partial differently for each of the images. I figured it would be more consistent to let the rendering templates to pass in a locals hash. What do you think?

wilsonyqm replied:
sounds good

wilsonyqm commented on app/controllers/images_controller.rb:
```diff
@@ -55,7 +55,15 @@ def edit
 
   def update
-    # TODO
+    @image = Image.find(params[:id])
```
better to use `find_by` here as well

wilsonyqm commented on app/controllers/images_controller.rb:
```diff
@@ -50,6 +50,22 @@ def destroy
+      flash[:success] = "Successfully updated image tags to: #{@image.tag_list}"
+      redirect_to images_path
+    else
+      render 'edit', status: :unprocessable_entity
```
it's better to set the flash[:error] as well here

wilsonyqm commented on test/controllers/image_controller_test.rb:
```diff
@@ -144,4 +144,48 @@ class ImageControllerTest < ActionDispatch::IntegrationTest
+    img = Image.create! url: 'http://test1.com/img1.png', tag_list: 'cat,tag2'
+    new_tags = 'generic, nonexistent, image'
+
+    patch image_path(img), params: { image: { tag_list: new_tags } }
```
why you use `patch` here, can you explain? we usually use `get :edit` or `post :update`

almgong replied:
I used the same HTTP method that `rails routes` showed for `images#update`. Turns out rails accepts both `PUT` and `PATCH`, and after looking online about the both, `PATCH` seemed to be more appropriate. From what I read, `PUT` usually involves sending the entire object to update (effectively replacing the data with what is sent in the request). `PATCH` seems to be used to identify changes to just parts of some data. However, I am sure rails understands our intent regardless of which of these methods are used. I have yet to see something like `get :edit` used to submit changes. How would that be done?

wilsonyqm replied:
`PATCH` is for Rails 4 specifically, `PUT/PATCH` is basically the same, here is what they said on stackoverflow:
```
A request using the POST method should act upon the resource collection; adding a new resource to the collection Example URL: http://example.com/resources

A request using the PUT HTTP verb should act upon a single resource within the collection; replacing the resource wholly upon the server Example URL: http://example.com/resource/1
```
Since you are updating on single tag here, so it makes sense if you use `patch`, nice job. In APM, we used `put` and I have never seen someone uses `patch`, it's good to know.

wilsonyqm commented on test/controllers/image_controller_test.rb:
```diff
@@ -144,4 +144,48 @@ class ImageControllerTest < ActionDispatch::IntegrationTest
+    patch image_path(img), params: { image: { tag_list: new_tags } }
+
+    assert_response :unprocessable_entity
+    assert_select '.error', 'must have at least one tag'
```
can you assert your flash[:danger] here as well?

wilsonyqm replied:
BTW, you probably have flash message and model validation error at the same time, right?

almgong replied:
Yeah, both the model validation (or form errors) and the flash message are shown at the same time. Do you think just one is enough, or should I keep it this way?

wilsonyqm replied:
I think the validation form errors should be good enough, otherwise it would be a bad UX design.

almgong replied:
Ok I will remove the flash altogether and just have the validation form errors

wilsonyqm replied:
Since you already have the `js` selector, we can use that selector here, `.js-edit-image .error`, this would be more specific, just in case if you have any other forms.

almgong replied:
I like the idea of making the `.error` more specific. Since the `.js-edit-image` is only applied to the actual image element, I will add a class to the form: `.js-edit-form`, and will then use `.js-edit-form .error` here.

wilsonyqm commented on test/page_objects/images/edit_page.rb:
```diff
@@ -0,0 +1,28 @@
+      end
+
+      def tags
+        tag_list.value.split(', ')
```
I noticed in some of your tests, the `tag_list` is not separated by `, ` with a whitespace, that probably would cause some issues for future implementation, it's better to use the same pattern every time or deal with specifically(like trim the white spaces).

almgong replied:
Good point. I will strip leading and trailing whitespace.

wilsonyqm commented on test/page_objects/images/edit_page.rb:
```diff
@@ -0,0 +1,28 @@
+      path :edit_image
+      path :image
+
+      element :img, locator: 'img'
```
One thing to notice, in APM, we use the locator with a specific name starts with `js`, for example, this locator here could be `.js-edit-image`, this class does not have any css styles with so it would be easy for us to tell the difference, it's been used a lot in selenium test.

almgong replied:
Sure, I will add a class `.js-edit-image` to the image.

wilsonyqm commented on test/flows/images_crud_test.rb:
```diff
@@ -88,4 +88,42 @@ class ImagesCrudTest < FlowTestCase
+
+    edit_image_page = images_show_page.click_edit!.as_a(PageObjects::Images::EditPage)
+
+    assert edit_image_page.image_url == puppy_url
```
we tried to get rid of assertion directly with `assert`, there is no error stacktrace in this case. I recommend you to use `assert_equal`, and place the `expected` content in the front, ex:
```ruby
assert_equal expected_content, result
```

wilsonyqm commented on test/flows/images_crud_test.rb:
```diff
@@ -88,4 +88,42 @@ class ImagesCrudTest < FlowTestCase
+    image_to_edit = images_index_page.images[0]
+    edit_image_page = image_to_edit.click_edit!.as_a(PageObjects::Images::EditPage)
+
+    assert edit_image_page.image_url == puppy_url
```
same as https://github.com/AppFolioOnboarding/image-sharer-almgong/pull/36/files/94f71006798d0d46a60f5543d650ff95e21d311a#r112788146

wilsonyqm commented:
Just something to think about, what if the tag is `,,,,`, what will happen, is your validation still valid?.

almgong commented:
>Just something to think about, what if the tag is ,,,,, what will happen, is your validation still valid?. How you want to prevent our customer entering some really bad data.

Yes, the validation will fail and the user will see the `must have at least one tag` form error. This is because the gem we use is intelligent enough to realize that there are no tags in something like `,,,`. They probably split the input string and check that the length of the resultant array is greater than 0.  But anyways, the case is handled :)

wilsonyqm commented on test/page_objects/images/edit_page.rb:
```diff
@@ -0,0 +1,28 @@
+      end
+
+      def tags
+        tag_list.value.split(', ').map(&:strip)
```
You don't need the whitespace here

almgong replied:
Good catch, removed.

wilsonyqm commented on app/views/images/_image_shared_actions.html.erb:
```diff
@@ -0,0 +1,2 @@
+<a href="#" class='btn btn-secondary' data-toggle="modal" data-target="#shareModal" data-form-action="<%= image_share_index_path(image.id) %>" data-image-url="<%= image.url %>">Share!</a>
```
Can you organize the lines to make it look better?

wilsonyqm commented on app/views/images/edit.html.erb:
```diff
@@ -0,0 +1,21 @@
+<div class="row">
```
can we try to use a partial name for the `edit`? like `_edit`, and call `render :partial` in the controller.

wilsonyqm replied:
That's how we usually do for rails default actions.

almgong replied:
Could your clarify this a bit? Do you mean to rename this file to something like `_edit.html.erb` and then in the controller always doing `render :partial`? I can't seem to find an example of this online

wilsonyqm replied:
Right, change the name to be `_edit` would be better, and use `render partial: 'edit'`

almgong replied:
I can't seem to get this to work. I changed the filename to `_edit.html.erb` and when I try to do: `render action: :edit` in the controller, I get the following error:

```
Missing template images/edit, application/edit with {:locale=>[:en], :formats=>[:html], :variants=>[], :handlers=>[:raw, :erb, :html, :builder, :ruby, :jbuilder]}. Searched in: * "/Users/appfolio/work/onboarding/image-sharer-almgong/app/views"
```

The only way I got it to work was if I did something like `render action: :_edit`. But this doesn't seem to be any better than just leaving the `edit` action blank, and letting rails automatically search for the correct `edit.html.erb` template. What exactly is the reasoning for rendering a partial vs a template in the `edit` action?

wilsonyqm commented on test/controllers/image_controller_test.rb:
```diff
@@ -144,4 +144,48 @@ class ImageControllerTest < ActionDispatch::IntegrationTest
+    get edit_image_path(img)
+
+    assert_response :ok
+    assert_select 'form[action=?]', image_path(img)
```
You can use `assert_template` here, it's kind of weird to use `assert_select`.

almgong replied:
I used the `assert_select` to check that there exists a form that has the expected form action. However, I guess checking that the correct template was rendered might be sufficient as well. I can replace `assert_select` with an `assert_template`.

almgong replied:
So I am getting an error that: `Minitest::UnexpectedError: NoMethodError: assert_template has been extracted to a gem. To continue using it, add `gem 'rails-controller-testing'` to your Gemfile.` Do you think I should install this gem just for this case, or continue with `assert_select` for now?

wilsonyqm replied:
Ohh, if we don't have that gem, using this `assert_select` is fine, you can specify the count of `assert_select` by doing `assert_select '.js-class', count: 1`, by default, the count is equal or larger than 1.

almgong replied:
Sounds, good I will add that count argument

wilsonyqm commented:
Very minor comments left. :)

wilsonyqm commented:
After addressing the comments, can you try to organize your commits better?

almgong commented:
Yeah, I have already squashed the previous commits. I got feedback from a different reviewer mentioning how difficult it was to find the changes I made (since I squashed them as I addressed each comment). That's why I have the weird `(f <commit hash>) ...` style. Basically I am saying where I will be squashing the commit into. But from now onwards for this PR, I can just squash them right into the commits they belong in, since the changes are smaller.

wilsonyqm commented:
Looks good!
